File: app/src/main/java/com/bracketcove/postrainer/usecase/CancelAlarm.java
package com.bracketcove.postrainer.usecase;


import com.bracketcove.postrainer.data.alarmservice.AlarmManager;
import com.bracketcove.postrainer.data.alarmservice.AlarmService;
import com.bracketcove.postrainer.data.viewmodel.Alarm;

import io.reactivex.Completable;

/**
 * Created by R_KAY on 5/23/2017.
 */

public class CancelAlarm implements UseCaseCompletable<Alarm> {

    private final AlarmManager alarmManager;

    public CancelAlarm(AlarmManager alarmManager) {
        this.alarmManager = alarmManager;
    }

    @Override
    public Completable runUseCase(Alarm... params) {
        return alarmManager.cancelAlarm(params[0]);
    }
}

File: domain/src/main/java/com/wiseassblog/domain/usecase/CancelAlarm.kt
package com.wiseassblog.domain.usecase


import com.wiseassblog.common.ResultWrapper
import com.wiseassblog.domain.domainmodel.Alarm
import com.wiseassblog.domain.repository.IAlarmRepository
import com.wiseassblog.domain.repository.IAlarmService

/**
 * 1. Tell alarmService to cancel the alarm
 * 2. Write alarm to database, with active = false
 *
 * Created by R_KAY on 5/23/2017.
 */

class CancelAlarm(
    private val service: IAlarmService,
    private val repo: IAlarmRepository
) : UseCaseWithParams<Alarm, ResultWrapper<Exception, Unit>>() {
    override suspend fun buildUseCase(params: Alarm): ResultWrapper<Exception, Unit> {
        val cancelResult = service.cancelAlarm(params)

        return if (cancelResult is ResultWrapper.Success) repo.updateAlarm(params.copy(isActive = false))
        else cancelResult
    }

}
